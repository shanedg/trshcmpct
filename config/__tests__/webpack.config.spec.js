const CleanWebpackPlugin = require('clean-webpack-plugin').CleanWebpackPlugin;
const HtmlWebpackPlugin = require('html-webpack-plugin');
const path = require('path');
const serializer = require('jest-serializer-path');

const webpackConfig = require('../webpack.config');

// Remove absolute paths from snapshots.
expect.addSnapshotSerializer(serializer);

describe('webpack', () => {
  const developmentEnvironment = {};
  const productionEnvironment = { production: true };
  let developmentConfig;
  let productionConfig;

  describe('common config', () => {

    beforeAll(() => {
      developmentConfig = webpackConfig(developmentEnvironment);
      productionConfig = webpackConfig(productionEnvironment);
      Object.freeze(developmentConfig);
      Object.freeze(productionConfig);
    });

    it('sets entry points', () => {
      expect(developmentConfig.entry).toHaveProperty('index', path.resolve(__dirname, '../../src/index.ts'));
      expect(productionConfig.entry).toStrictEqual(developmentConfig.entry);
    });

    it('sets output', () => {
      expect(developmentConfig.output).toHaveProperty('filename', '[name].[chunkhash].js');
      expect(developmentConfig.output).toHaveProperty('path', path.resolve(__dirname, '../../dist'));
      expect(productionConfig.output).toStrictEqual(developmentConfig.output);
    });

    it('sets module rules', () => {
      expect(developmentConfig.module).toHaveProperty('rules', expect.any(Array));
      expect(productionConfig.module).toStrictEqual(developmentConfig.module);
    });

    it('sets optimizations', () => {
      expect(developmentConfig.optimization).toHaveProperty('splitChunks', expect.any(Object));
      expect(productionConfig.optimization).toStrictEqual(developmentConfig.optimization);
    });

    it('sets plugins', () => {
      expect(developmentConfig.plugins).toMatchObject([
        expect.any(CleanWebpackPlugin),
        expect.any(Function), // ProgressBarPlugin, anonymous function signature
        expect.any(HtmlWebpackPlugin),
      ]);
      // Can't use deep-equality checks around plugins since they're unique class instances so repeat same assertion against productionConfig.
      // Otherwise, toEqual/toStrictEqual comparisons fail with "Received: serializes to the same string".
      expect(productionConfig.plugins).toMatchObject([
        expect.any(CleanWebpackPlugin),
        expect.any(Function), // ProgressBarPlugin
        expect.any(HtmlWebpackPlugin),
      ]);
    });

    it('sets dev server options', () => {
      expect(developmentConfig.devServer).toHaveProperty('open', true);
      expect(productionConfig.devServer).toStrictEqual(developmentConfig.devServer);
    });

  });

  describe('development config', () => {

    beforeAll(() => {
      developmentConfig = webpackConfig(developmentEnvironment);
      Object.freeze(developmentConfig);
    });

    it('matches snapshot', () => {
      expect(developmentConfig).toMatchSnapshot();
    });

    it('sets development mode', () => {
      expect(developmentConfig.mode).toEqual('development');
    });

  });

  describe('production config', () => {

    beforeAll(() => {
      productionConfig = webpackConfig(productionEnvironment);
      Object.freeze(productionConfig);
    });

    it('matches snapshot', () => {
      expect(productionConfig).toMatchSnapshot();
    });

    it('sets production mode', () => {
      expect(productionConfig.mode).toEqual('production');
    });

  });

});
