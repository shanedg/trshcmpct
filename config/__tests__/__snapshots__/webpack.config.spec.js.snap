// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webpack development config matches snapshot 1`] = `
Object {
  "devServer": Object {
    "open": true,
  },
  "entry": Object {
    "index": "<PROJECT_ROOT>/src/index.ts",
  },
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "enforce": "pre",
        "exclude": /node_modules/,
        "loader": "eslint-loader",
        "options": Object {
          "cache": true,
          "emitError": false,
          "emitWarning": true,
        },
        "test": /\\\\\\.\\(js\\|jsx\\|ts\\|tsx\\)\\$/,
      },
      Object {
        "exclude": /node_modules/,
        "loader": "babel-loader",
        "options": Object {
          "configFile": "<PROJECT_ROOT>/config/babel.config.js",
        },
        "test": /\\\\\\.\\(js\\|jsx\\|ts\\|tsx\\)\\$/,
      },
    ],
  },
  "optimization": Object {
    "splitChunks": Object {
      "cacheGroups": Object {
        "vendors": Object {
          "priority": -10,
          "test": /\\[\\\\\\\\\\\\/\\]node_modules\\[\\\\\\\\\\\\/\\]/,
        },
      },
      "chunks": "async",
      "minChunks": 1,
      "minSize": 30000,
      "name": true,
    },
  },
  "output": Object {
    "filename": "[name].[chunkhash].js",
    "path": "<PROJECT_ROOT>/dist",
  },
  "plugins": Array [
    CleanWebpackPlugin {
      "apply": [Function],
      "cleanAfterEveryBuildPatterns": Array [],
      "cleanOnceBeforeBuildPatterns": Array [
        "**/*",
      ],
      "cleanStaleWebpackAssets": true,
      "currentAssets": Array [],
      "dangerouslyAllowCleanPatternsOutsideProject": false,
      "dry": false,
      "handleDone": [Function],
      "handleInitial": [Function],
      "initialClean": false,
      "outputPath": "",
      "protectWebpackAssets": true,
      "removeFiles": [Function],
      "verbose": false,
    },
    [Function],
    DefinePlugin {
      "definitions": Object {
        "__DEV__": "true",
      },
    },
    HtmlWebpackPlugin {
      "options": Object {
        "cache": true,
        "chunks": "all",
        "chunksSortMode": "auto",
        "compile": true,
        "excludeChunks": Array [],
        "favicon": false,
        "filename": "index.html",
        "hash": false,
        "inject": true,
        "meta": Object {},
        "minify": false,
        "showErrors": true,
        "template": "<PROJECT_ROOT>/src/index.html",
        "templateParameters": [Function],
        "title": "trshcmpctr",
        "xhtml": false,
      },
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".wasm",
      ".mjs",
      ".js",
      ".json",
      ".jsx",
      ".ts",
      ".tsx",
    ],
  },
}
`;

exports[`webpack production config matches snapshot 1`] = `
Object {
  "devServer": Object {
    "open": true,
  },
  "entry": Object {
    "index": "<PROJECT_ROOT>/src/index.ts",
  },
  "mode": "production",
  "module": Object {
    "rules": Array [
      Object {
        "enforce": "pre",
        "exclude": /node_modules/,
        "loader": "eslint-loader",
        "options": Object {
          "cache": true,
          "emitError": true,
          "emitWarning": false,
        },
        "test": /\\\\\\.\\(js\\|jsx\\|ts\\|tsx\\)\\$/,
      },
      Object {
        "exclude": /node_modules/,
        "loader": "babel-loader",
        "options": Object {
          "configFile": "<PROJECT_ROOT>/config/babel.config.js",
        },
        "test": /\\\\\\.\\(js\\|jsx\\|ts\\|tsx\\)\\$/,
      },
    ],
  },
  "optimization": Object {
    "splitChunks": Object {
      "cacheGroups": Object {
        "vendors": Object {
          "priority": -10,
          "test": /\\[\\\\\\\\\\\\/\\]node_modules\\[\\\\\\\\\\\\/\\]/,
        },
      },
      "chunks": "async",
      "minChunks": 1,
      "minSize": 30000,
      "name": true,
    },
  },
  "output": Object {
    "filename": "[name].[chunkhash].js",
    "path": "<PROJECT_ROOT>/dist",
  },
  "plugins": Array [
    CleanWebpackPlugin {
      "apply": [Function],
      "cleanAfterEveryBuildPatterns": Array [],
      "cleanOnceBeforeBuildPatterns": Array [
        "**/*",
      ],
      "cleanStaleWebpackAssets": true,
      "currentAssets": Array [],
      "dangerouslyAllowCleanPatternsOutsideProject": false,
      "dry": false,
      "handleDone": [Function],
      "handleInitial": [Function],
      "initialClean": false,
      "outputPath": "",
      "protectWebpackAssets": true,
      "removeFiles": [Function],
      "verbose": false,
    },
    [Function],
    DefinePlugin {
      "definitions": Object {
        "__DEV__": "false",
      },
    },
    HtmlWebpackPlugin {
      "options": Object {
        "cache": true,
        "chunks": "all",
        "chunksSortMode": "auto",
        "compile": true,
        "excludeChunks": Array [],
        "favicon": false,
        "filename": "index.html",
        "hash": false,
        "inject": true,
        "meta": Object {},
        "minify": false,
        "showErrors": true,
        "template": "<PROJECT_ROOT>/src/index.html",
        "templateParameters": [Function],
        "title": "trshcmpctr",
        "xhtml": false,
      },
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".wasm",
      ".mjs",
      ".js",
      ".json",
      ".jsx",
      ".ts",
      ".tsx",
    ],
  },
}
`;
